lib/api/apiUtils/object/coldStorage.js

```diff
@@ -28,6 +31,131 @@ function getAmzRestoreResHeader(objMD) {
+        && new Date(objectMD.archive?.restoreWillExpireAt) < new Date(Date.now())) {
+        // return InvalidObjectState error if the restored object is expired
+        // but restore info md of this object has not yet been cleared
+        log.debug('The restored object already expired.',
+            {
+                archive: objectMD.archive,
+                method: '_validateStartRestore',
+            });
+        return errors.InvalidObjectState;
+    }
+    if (objectMD.archive?.restoreRequestedAt && !objectMD.archive?.restoreCompletedAt) {
+        // return RestoreAlreadyInProgress error if the object is currently being restored
+        // check if archive.restoreRequestAt exists and archive.restoreCompletedAt not yet exists
+        log.debug('The object is currently being restored.',
+            {
+                archive: objectMD.archive,
+                method: '_validateStartRestore',
+            });
+        return errors.RestoreAlreadyInProgress;
+    }
+    return undefined;
+}
+
+/**
+ * Check if the object is already restored
+ *
+ * @param {ObjectMD} objectMD - object metadata
+ * @param {object} log - werelogs logger
+ * @return {boolean} - true if the object is already restored
+ */
+function isObjectAlreadyRestored(objectMD, log) {
+    //  check if restoreCompletedAt field exists
+    //  and archive.restoreWillExpireAt > current time
+    const isObjectAlreadyRestored = objectMD.archive?.restoreCompletedAt
+        && new Date(objectMD.archive?.restoreWillExpireAt) >= new Date(Date.now());
+    log.info(`Object is ${isObjectAlreadyRestored ? '' : 'not'} already restored.`,
+        {
+            archive: objectMD.archive,
+            method: 'isObjectAlreadyRestored'
+        });
+    return isObjectAlreadyRestored;
+}
+
+/**
+ * update restore expiration date.
+ *
+ * @param {ObjectMD} objectMD - objectMD instance
+ * @param {object} restoreParam - restore param
+ * @param {object} log - werelogs logger
+ * @return {ArsenalError|undefined} internal error if object MD is not valid
+ *
+ */
+function _updateRestoreInfo(objectMD, restoreParam, log) {
+    if (!objectMD.archive) {
+        log.debug('objectMD.archive doesn\'t exits', {
+            objectMD,
+            method: '_updateRestoreInfo'
+        });
+        return errors.InternalError.customizeDescription('objectMD.archive doesn\'t exit for an archived object.');
+    }
+    // eslint-disable-next-line no-param-reassign
+    objectMD.archive.restoreRequestedAt = new Date();
+    // eslint-disable-next-line no-param-reassign
+    objectMD.archive.restoreRequestedDays = restoreParam.days;
+    if (!ObjectMDArchive.isValid(objectMD.archive)) {
+        log.debug('archive is not valid', {
+            archive: objectMD.archive,
+            method: '_updateRestoreInfo'
+        });
+        return errors.InternalError.customizeDescription('Invalid archive object. ' +
+            'Please check ObjectMDArchive validation function.');
```

Comment:
Same here, the end-user of the S3 service is certainly not going to check the validation function. And the log should not be debug, but error, to help find the error cause.

If it's a serious invariant that is violated, an assert+crash would even be suitable (but not the case here).